plan/views.py

```diff
@@ -32,3 +39,83 @@ def get_queryset(self):
 
 class CourseDetailSearch(CourseDetail):
     serializer_class = CourseDetailWithReviewSerializer
+
+
+def get_sections(data):
+    sections = None
+    if 'meetings' in data:
+        sections = []
+        for s in data.get('meetings'):
+            _, section = get_course_and_section(s.get('id'), s.get('semester'))
+            sections.append(section)
+    elif 'sections' in data:
+        sections = []
+        for s in data.get('sections'):
+            _, section = get_course_and_section(s.get('id'), s.get('semester'))
+            sections.append(section)
+    return sections
+
+
+class ScheduleViewSet(viewsets.ModelViewSet):
+    serializer_class = ScheduleSerializer
+    http_method_names = ['get', 'post', 'delete', 'put']
+    permission_classes = [IsAuthenticated]
+
+    def update(self, request, pk=None):
+        existing_obs = Schedule.objects.filter(id=pk)
+        if (len(existing_obs) == 0):
+            return Response({'detail': 'Not found.'}, status=status.HTTP_404_NOT_FOUND)
+
+        if 'semester' not in request.data:
+            request.data['semester'] = get_value('SEMESTER', None)
+
+        for s in request.data['sections']:
+            if s.get('semester') != request.data.get('semester'):
+                return Response({'detail': 'Semester uniformity invariant violated.'},
+                                status=status.HTTP_400_BAD_REQUEST)
+
+        try:
+            existing_obs.update(person=request.user,
+                                semester=request.data.get('semester'),
+                                name=request.data.get('name'),
+                                )
+            ob = existing_obs[0]
+            ob.sections.set(get_sections(request.data))
+            s = ScheduleSerializer(ob)
+            return Response(s.data, status=status.HTTP_202_ACCEPTED)
+        except IntegrityError:
+            return Response({'detail': 'Unique constraint violated'}, status=status.HTTP_400_BAD_REQUEST)
+
+    def create(self, request, *args, **kwargs):
+        existing_obs = Schedule.objects.filter(id=request.data.get('id'))
+        if len(existing_obs) > 0:
+            return self.update(request, request.data.get('id'))
+
+        if 'semester' not in request.data:
+            request.data['semester'] = get_value('SEMESTER', None)
+
+        for s in request.data['sections']:
+            if s.get('semester') != request.data.get('semester'):
+                return Response({'detail': 'Semester uniformity invariant violated.'},
+                                status=status.HTTP_400_BAD_REQUEST)
+
+        try:
+            if 'id' in request.data:
```

Comment:
The correct behavior is probably to ignore the given id when creating a schedule, since that might cause a collision with an existing schedule that we wouldn't want to happen.